File: src/main/java/kitchenpos/menus/domain/tobe/MenuPrice.java
package kitchenpos.menus.domain.tobe;

import jakarta.persistence.Column;
import jakarta.persistence.Embeddable;
import java.math.BigDecimal;
import java.util.Objects;

@Embeddable
public class MenuPrice {

    @Column(name = "price", nullable = false)
    private BigDecimal price;

    protected MenuPrice() {
    }

    public MenuPrice(BigDecimal price) {
        if (Objects.isNull(price) || price.compareTo(BigDecimal.ZERO) < 0) {
            throw new IllegalArgumentException();
        }

        this.price = price;
    }

    public boolean isOver(BigDecimal price) {
        return this.price.compareTo(price) > 0;
    }

    public BigDecimal getPrice() {
        return price;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        MenuPrice menuPrice = (MenuPrice) o;
        return Objects.equals(price, menuPrice.price);
    }

    @Override
    public int hashCode() {
        return Objects.hash(price);
    }
}

File: src/main/kotlin/kitchenpos/tobe/menu/domain/vo/MenuPrice.kt
package kitchenpos.tobe.menu.domain.vo

import jakarta.persistence.Column
import jakarta.persistence.Embeddable
import java.math.BigDecimal

@Embeddable
class MenuPrice private constructor(
    @Column(name = "price", nullable = false)
    val price: BigDecimal,
) {
    companion object {
        fun from(
            price: BigDecimal,
            productsPrice: BigDecimal,
        ): MenuPrice {
            require(
                price <= productsPrice,
            ) {
                throw IllegalArgumentException("메뉴의 가격은 메뉴에 속한 상품의 가격 합보다 작거나 같아야 합니다.")
            }
            return MenuPrice(price)
        }
    }
}
